L82-core-event-engine.md

```diff
@@ -0,0 +1,538 @@
+```
+
+#### Setting an EventEngine instance for a Server
+
+A gRPC server can use a custom EventEngine by calling the
+`ServerBuilder::SetEventEngine` method:
+
+```cpp
+ServerBuilder builder;
+std::shared_ptr<EventEngine> engine = std::make_shared<MyEngine>(...);
+builder.SetEventEngine(engine);
+std::unique_ptr<Server> server(builder.BuildAndStart());
+server->Wait();
+```
+
+#### Providing a default EventEngine factory
+
+TODO: this has not been thoroughly discussed.
+
+Before gRPC is initialized, an application can specify an EventEngine Factory
+method that will be used within gRPC when a channel- or server-specific
+EventEngine is unavailable. gRPC needs to perform I/O, run tasks, and use timers
+in scenarios that are not directly attached to channels and servers.  Providing
+a custom EventEngine factory will allow your application to avoid the creation
+of the default EventEngine provided by gRPC.
+
+### Lifetime and Ownership
+
+gRPC will take shared ownership of application-provided EventEngines via
+std::shared\_ptrs to ensure that the engines remain available until they are no
+longer needed. Depending on the use case, EventEngines may need to live until
+gRPC is shut down. Any EventEngine that is provided to gRPC is expected to
+remain valid for as long as gRPC holds a pointer to it. Application-provided
+EventEngines will not be shut down automatically when gRPC is shut down, it is
+the application's responsibility to shut down and clean up EventEngines when
+they are no longer needed.
+
+### Temporary protections
+
+The EventEngine can be enabled at compile time by defining the C/C++ macro
+`GRPC_USE_EVENT_ENGINE`. EventEngines will not be enabled by default on any
+platform to begin with.
+
+The EventEngine API will exist in the `grpc_event_engine::experimental`
+namespace until it has stabilized. Stabilization will likely be signaled by
+having all platforms supported sufficiently by the default EventEngine, and
+having all other primary EventEngine use cases exercised without needing to make
+further API changes.
+
+The EventEngine rollout is planned to be on a per-platform basis. Once we
+establish that the default EventEngine implementation is sufficient for a
+specific platform, we will change the build to replace iomgr with EventEngine on
+that platform.
+
+## Limitations
+
+### No immediate goal for wrapped languages to provide custom EventEngines
+
+The ability to provide custom EventEngines at runtime will initially be
+available in the C++ and C-core APIs only. Adding support for custom
+EventEngines written in wrapped languages, or providing EventEngines at runtime
+from wrapped languages, is not a focus of this work.  Challenges may include the
+management of dll-barrier crossing issues, and potential performance losses.
+
+### Public file descriptor operations are not supported for custom EventEngines
+
+gRPC public APIs contain methods that revolve around the concept of a system
```

Comment:
We know that we have these methods in C++ and C-core.  Have we also audited wrapped languages to see if any of these APIs are used there?  If not, we should.